clastb representation in existing IR, and AArch64 codegen

huntergr-arm opened the pull request:
These commits show a possible representation of SVE's `clastb` instruction using existing IR instructions and intrinsics, along with DAGCombines to emit the actual instruction.

At 9 instructions to represent `clastb`, it feels a little fragile and may be changed by passes between LoopVectorize and codegen. While we can sink the loop-invariant terms back into the right block in CGP, I do wonder if we want a more direct intrinsic to represent this kind of operation.

Perhaps something like `llvm.vector.extract.last.active(data, mask)` ?

This is something we would use to support the CSA vectorization in #106560 for SVE, though we would prefer to use clastb inside the vector loop instead of after it. That patch uses an int max reduction to determine the index instead of the cttz.elts based approach in this PR, so we have another existing IR option to use if we want.

efriedma-quic commented:
Reliably matching the integer arithmetic seems problematic; I think I'd lean towards an intrinsic.   Maybe `llvm.vector.extract.last.active(data, mask, passthru)`, set the passthru to poison if you don't care.

huntergr-arm commented:
Rebased on top of the intrinsic.

SamTebbs33 reviewed (approved):
Looks good to me apart from a doc comment needed.

arsenm reviewed (changes requested):
Bad TargetLowering API with no use context. SelectionDAGBuilder should directly translate into dag nodes and not be a point of configuration

huntergr-arm commented:
> Bad TargetLowering API with no use context. SelectionDAGBuilder should directly translate into dag nodes and not be a point of configuration

I was following a pattern of existing TLI functions (starting with `shouldExpandGetActiveLaneMask` 3 years ago). If that's the wrong way to do it, then I guess we have some cleanup to do later in addition to removing the function from this patch.

I'll investigate leaving it as an intrinsic and moving the current code in SelectionDAGBuilder to an expand legalization function instead.

arsenm commented:
> I was following a pattern of existing TLI functions (starting with `shouldExpandGetActiveLaneMask` 3 years ago). If that's the wrong way to do it, then I guess we have some cleanup to do later in addition to removing the function from this patch.
> 

It's developing an ad-hoc, parallel legalization mechanism to the regular one that happens in a different place. This expands the surface area of developing a target, and I have endless trouble fighting with non-trivial logic in the builder. All the legalization decisions should be made during legalization, not before

huntergr-arm committed:
[AArch64] Basic ISel for find_last_active

huntergr-arm committed:
Combine extract_vector_elt(find_last_active) -> lastb

huntergr-arm committed:
Combine csel+lastb -> clastb

huntergr-arm committed:
Formatting

huntergr-arm commented:
Rebased on the vecops-expansion-based lowering from #118810. I implemented basic SVE isel pattern matching for the new ISD node, then added a couple new dag combines to fold to lastb, then clastb.

huntergr-arm commented:
@arsenm has the changed design addressed your concerns? The SDAGBuilder lowering is the same for all targets, and now I'm just marking the new ISD node as Legal for AArch64 to avoid the default vecops expansion here to improve SVE codegen.

MacDue commented on llvm/include/llvm/Target/TargetSelectionDAG.td:
```diff
@@ -840,6 +840,9 @@ def vector_insert_subvec : SDNode<"ISD::INSERT_SUBVECTOR",
 def insert_subvector : SDNode<"ISD::INSERT_SUBVECTOR", SDTSubVecInsert, []>;
 
+def find_last_active
+    : SDNode<"ISD::VECTOR_FIND_LAST_ACTIVE", SDTypeProfile<1, 1, []>, []>;
+
```
Is it worth adding any constrains for `find_last_active`?

MacDue reviewed (approved):
LGTM, I think @arsenm's main concern as been addressed as there's no new target lowering hook anymore. I think the PR description needs updating though as the `llvm.vector.extract.last.active` intrinsic now exists and this patch is improves the SVE lowering.

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64SVEInstrInfo.td:
```diff
@@ -3379,6 +3379,20 @@ let Predicates = [HasSVE_or_SME] in {
   def : Pat<(i64 (vector_extract nxv2i64:$vec, VectorIndexD:$index)),
             (UMOVvi64 (v2i64 (EXTRACT_SUBREG ZPR:$vec, zsub)), VectorIndexD:$index)>;
 
+  // Find index of last active lane. This is a fallback in case we miss the
+  // opportunity to fold into a lastb or clastb directly.
```
I agree it would be good to have some tests for these.

huntergr-arm replied:
Sadly it's pretty difficult to do this once the combines have been added. I don't see a global switch to disable combining, just the target-indepedent combines. We do check the optimization level in a few places in AArch64ISelLowering, but mostly for TLI methods for IR-level decisions. Deliberately turning off (c)lastb pattern matching at O0 feels odd. Adding a new switch just for this feature also feels excessive.

I could potentially add a globalisel-based test, though I'm not sure how much code that requires. We've added a few new ISD nodes recently, and none have added support in globalisel.

I guess this is mostly due to it being hard to just create a selectiondag without IR and run selection over it.

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -19944,6 +19945,28 @@ performLastTrueTestVectorCombine(SDNode *N,
+  SDValue Vec = N->getOperand(0);
+  SDValue Idx = N->getOperand(1);
+
+  if (!Subtarget->hasSVE() || DCI.isBeforeLegalize() ||
```
nit: the test for `Subtarget->hasSVE()` is redundant and tested by the `getOperationAction(..) == legal` below.

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -25032,6 +25057,41 @@ static SDValue reassociateCSELOperandsForCSE(SDNode *N, SelectionDAG &DAG) {
+  if (LastB.getOpcode() != AArch64ISD::LASTB || LastB.getOperand(0) != AnyPred)
+    return SDValue();
+
+  SDValue Vec = LastB.getOperand(1);
```
nit: only has one use, can just as well propagate into expression below.

huntergr-arm committed:
Type constraints for find last active ISD node

arsenm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -19944,6 +19945,28 @@ performLastTrueTestVectorCombine(SDNode *N,
+
+  SDValue Mask = Idx.getOperand(0);
+  const TargetLowering &TLI = DAG.getTargetLoweringInfo();
+  if (TLI.getOperationAction(ISD::VECTOR_FIND_LAST_ACTIVE,
+                             Mask.getValueType()) != TargetLowering::Legal)
```
isOperationLegal?

huntergr-arm committed:
Restrict combine by element types

huntergr-arm committed:
Don't flag nxv1i1 as legal for extract last

huntergr-arm committed:
Use isOperationLegal

huntergr-arm committed:
Remove single-use variable

huntergr-arm committed:
Formatting

huntergr-arm committed:
Initial half/bfloat tests

huntergr-arm committed:
Perform last active DAG combine on half/bfloat scalable vectors
